File: translated/f-to-c/DRB109-orderedmissing-orig-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//This is a program based on a test contributed by Yizi Gu@Rice Univ.
/* Missing the ordered clause
 * Data race pair: x@21:9:W vs. x@21:9:W
 */

#include <omp.h>
#include <stdio.h>

int main()
{
    int x, i;
    x = 0;

    #pragma omp parallel for ordered
    for (i = 0; i < 100; i++) {
        // Missing ordered directive here causes data race
        x = x + 1;
    }

    printf("x = %d\n", x);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB109-orderedmissing-orig-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//This is a program based on a test contributed by Yizi Gu@Rice Univ.
// * Missing the ordered clause
// * Data race pair: x@21:9:W vs. x@21:9:W

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func main() {
	var x, i int

	//$omp parallel do ordered
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := 100 / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= 100; start += chunkSize {
		end := start + chunkSize - 1
		if end > 100 {
			end = 100
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				x = x + 1
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("x = %d\n", x)
}
